Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,330 @@
+
+/--
+Swaps two entries in an vector, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swap! (v : Vector α n) (i j : Nat) : Vector α n :=
+  ⟨Array.swap! v.toArray i j, by {rw [size_swap!]; exact v.size_eq}⟩
+
+/--
+Swaps the entry with index `i : Fin n` in the vector for a new entry.
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapAt (v : Vector α n) (i : Fin n) (x : α) : α × Vector α n:=
+  let e := Vector.get v i
+  let v := Vector.set v i x
+  (e, v)
+
+instance [Inhabited α] : Inhabited (Vector α n) where
+  default := mkVector n default
+
+/--
+`swapAt! v i x` swaps out the entry at index `i : Nat` in the vector for `x`, if the index is valid. Otherwise it panics The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapAt! (v : Vector α n) (i : Nat) (x : α) : α × Vector α n :=
+  if h : i < n then
+    swapAt v ⟨i, h⟩ x
+  else
+    have : Inhabited α := ⟨x⟩
+    panic! ("index " ++ toString i ++ " out of bounds")
+
+/-- `range n` Returns a vector `#v[1,2,3,...,n-1]` -/
+def range (n : Nat) : Vector Nat n :=
+  ⟨Array.range n, by {simp}⟩
+
+/-- Returns a vector of size `1` with a single element `v` -/
+def singleton (v : α) : Vector α 1 :=
+  mkVector 1 v
+
+/-- Vector lookup function that takes an index `i` of type `USize` -/
+def uget (v : Vector α n) (i : USize) (h : i.toNat < n) : α :=
+  Array.uget v.toArray i (v.size_eq.symm ▸ h)
+
+instance : GetElem (Vector α n) Nat α fun _ i => i < n where
+  getElem := fun x i h => get x ⟨i, h⟩
+
+/-- `pop v` returns the vector with the last element removed -/
+def pop (v : Vector α n) : Vector α (n-1) :=
+  ⟨Array.pop v.toArray, by {rw [size_pop, v.size_eq]}⟩
+
+/-- `shrink v m` shrinks the vector to the first `m` elements if `m < n`. -/
+def shrink (v : Vector α n) (m : Nat) : Vector α (min m n) :=
+  ⟨Array.shrink v.toArray m, proof⟩
+  where
+    proof := by
+      rw [Array.shrink, Array.size_shrink_loop, v.size_eq]
+      omega
+
+/--
+`isEqv` takes a given boolean property `p`. It returns `true`
+if and only if `p a[i] b[i]` holds true for all valid indices `i`.
+-/
+def isEqv (a b : Vector α n) (p : α → α → Bool) : Bool :=
+  Array.isEqv a.toArray b.toArray p
+
+instance [BEq α] : BEq (Vector α n) :=
+  ⟨fun a b => isEqv a b BEq.beq⟩
+
+/-- `reverse v` reverses the vector `v` -/
+def reverse (v : Vector α n) : Vector α n :=
+  ⟨Array.reverse v.toArray, proof⟩
+  where
+    proof := by
+      rw [size_reverse]
+      exact v.size_eq
+/--
+Removes the element at a given index from a vector without bounds checks, using a Fin index.
+
+This function takes worst case O(n) time because it has to backshift all elements at positions
+greater than i.
```

Comment:
```suggestion
Removes the element at a given index from a vector using a Fin index.

This function takes worst case O(n) time because it has to backshift all elements at positions
greater than i.
```